File: compiler/bench/cse1.c
// expected: 1100000000
// Common subexpression elimination - repeated computation
#include <stdio.h>

static long compute(long a, long b) {
    return (a + b) + (a * b);
}

int main() {
    long sum = 0;
    for (long i = 0; i < 100000000; i++) {
        sum += compute(2, 3);
    }
    printf("%ld\n", sum);
    return 0;
}

File: compiler/bench/cse1.fs
\ expected: 1100000000
\ Common subexpression elimination - repeated computation

: compute ( a b -- result )
  over over + >r   \ a+b saved
  over over * >r   \ a*b saved
  2drop
  r> r> + ;        \ (a*b) + (a+b)

: main
  0 100000000 0 do
    2 3 compute +
  loop . cr ;
